### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by myitcv** [cue-lang/cue#2850]
Performance is a key goal of the CUE project. Broadly, @mpvl envisages being able to achieve performance improvements of several orders of magnitudes (predominantly by means of reductions in the time complexity).

Performance is also a key requirement of developer tooling, for example the CUE LSP.

This umbrella issue exists to summarize the main categories of current performance problems that people run into. It will link to sub-issues that capture more specific detail on well-identified categories of performance problems. We will also use this umbrella issue to post high-level updates on performance matters: it will therefore be locked to allow subscribing to updates from @mpvl. 

Please subscribe to this issue to receive notifications of those updates. 

The sub-issues linked below this umbrella issue will include more detail of each category of performance problem. Each sub-issue will link via [tasklists](https://docs.github.com/en/issues/managing-your-tasks-with-tasklists/about-tasklists) to bug reports of performance issues that are known to suffer from the category of performance problem. High-level updates specific to each category of performance problem posted to those interested in subscribing to sub-issues.

The main umbrella issue and sub-issues will be part of the [Evaluator Roadmap](https://github.com/orgs/cue-lang/projects/18) project (see our [recent announcement of new approach to issue/project management](https://github.com/cue-lang/cue/discussions/2846)). Bug reports linked from sub-issues will not be added to the Evaluator Project because they are entirely covered by the sub-issue.

The main categories of performance problems are:

* [Disjunctions](https://github.com/cue-lang/cue/issues/2851)
* [Repeated evaluation of incomplete expressions](https://github.com/cue-lang/cue/issues/2852)
* [Closedness algorithm consuming lots of memory](https://github.com/cue-lang/cue/issues/2853)
* [Avoiding duplicate work](https://github.com/cue-lang/cue/issues/2854)
* [Incremental evaluation](https://github.com/cue-lang/cue/issues/2855)

For discussion of this issue and all matters related to performance please see https://github.com/cue-lang/cue/discussions/2857. 

## FAQ

_**My CUE is slow to evaluate - what should I do?**_
Please [raise an issue](https://github.com/cue-lang/cue/issues/new/choose) and we will help to investigate. Part of that investigation will involve working out which of the categories of performance problems affect your code, and linking to your issue from a tasklist in the relevant sub-issue.

_**Where can we discuss things relating to this umbrella issue?**_

For discussion about performance issues, please see https://github.com/cue-lang/cue/discussions/2857, or create [a new discussion](https://github.com/cue-lang/cue/discussions/new/choose) and we will be happy to join the exchange.

_**Why is my performance bug not part of the Evaluator Project?**_

Where there are well-identified categories of performance problems, it makes sense to group these as sub-issues under the umbrella performance issue. This helps to keep the project board small, whilst not losing any detail in the original report. The linkage to the original bug reports is managed from sub-issues via tasklists.

**Comment by myitcv** [comment]
As a follow up to this issue and the nexts steps outlined in https://github.com/cue-lang/cue/discussions/2857, we are arranging an Evaluator and Performance Office Hours on Thursday 29 Feb, 2024 at 1600 UTC. For more details please see https://github.com/cue-lang/cue/discussions/2862.

**Comment by mpvl** [comment]
To update on the progress of the development of the new evaluator, a critical component of making performance improvements in CUE.

The [Evaluator Roadmap project](https://github.com/orgs/cue-lang/projects/18) is gradually being filled out with issues to communicate our progress in this part of the CUE project. For more information on the approach we are following here, please see https://github.com/cue-lang/cue/discussions/2846. 

Back in December 2023 on our [last community call](https://github.com/cue-lang/cue/discussions/2724), we updated to confirm the new evaluator was close to passing all tests except for disjunctions. The changes to support this work (i.e. everything except disjunctions) have since landed in the main CUE repo. We will be submitting the [final key change](https://review.gerrithub.io/c/cue-lang/cue/+/1174341) in this series soon.

Since December, I have been making good progress with disjunctions. We originally planned to have a simple implementation for disjunctions to speed up the roll out. That turned out to be problematic. The latest implementation (for the new evaluator), although still conceptually simpler than the previous one, took a bit more effort than anticipated. Things are going well now, though. 

We now plan to submit the changes to support disjunctions in the new evaluator in an incremental fashion.

### Sequencing

A bit more about next steps.

https://github.com/cue-lang/cue/issues/2853 is largely done with the exception of reclaiming memory (see https://github.com/cue-lang/cue/issues/2887)  and some esoteric bugs. 

We first need to land https://review.gerrithub.io/c/cue-lang/cue/+/1174341, the main change implementing the new evaluator. This is being tracked in https://github.com/cue-lang/cue/issues/2884. 

Once the biggest bugs with disjunctions are ironed out (which we are doing via tests and code that is part of Unity), we plan to allow users to play with the new evaluator, enabling it via a `CUE_EXPERIMENT`. This is being tracked in https://github.com/cue-lang/cue/issues/2886. 

Note that this initial experimental version of the new evaluator will not include memory reuse or any of the secondary optimizations, such as structure sharing (see https://github.com/cue-lang/cue/issues/2854). So even though users should see an improvement in the big-O performance of the core evaluator, it might be that we initially see performance get worse in some instances. With such a significant change in implementation, this is somewhat to be expected because the implementations of the evaluator (old/current and new) have very different designs and therefore characteristics.  

Once support for enabling the `CUE_EXPERIMENT` is released, the next step is to enable reuse buffers. We have disabled memory reclamation as it simplifies, and thus speeds up, development. Enabling it will give a critical boost. For instance, the old closedness algorithm would accumulate oodles of auxiliary structures that could not be reclaimed (see https://github.com/cue-lang/cue/issues/2853). The new evaluator allows reclaiming this memory after any small evaluation increment. Until this is enabled, however, users will not see much benefits of this new algorithm. This is tracked in https://github.com/cue-lang/cue/issues/2887. 

We also need to ensure that error messages in the new evaluator are at least on par with the old evaluator. This is tracked in https://github.com/cue-lang/cue/issues/2890. The new evaluator enables more structured capturing of error situations, and therefore better error reporting. Improving error messages is tracked via https://github.com/cue-lang/cue/issues/2891. 

After this is done, the real fun starts. The new evaluator has been designed with several additional performance improvements in mind. Each of those can give substantial constant speedup, or even additional big-O improvements. These can all be rolled out incrementally. We will be creating issues as part of the Evaluator Roadmap for each of these. 

### Updates on main performance sub-issues

This update being the first of its kind, means it necessarily includes background and context on where we find ourselves today. Future updates (which are aiming to be at least fortnightly) will be shorter. In future updates we will provide shorter bullets under the heading of the sub-issues:

-   [Disjunctions](https://github.com/cue-lang/cue/issues/2851)
-   [Repeated evaluation of incomplete expressions](https://github.com/cue-lang/cue/issues/2852)
-   [Closedness algorithm consuming lots of memory](https://github.com/cue-lang/cue/issues/2853)
-   [Avoiding duplicate work](https://github.com/cue-lang/cue/issues/2854)
-   [Incremental evaluation](https://github.com/cue-lang/cue/issues/2855)

### Key open changes

Here is a list of the active changes related to the evaluator and performance work, referenced either directly or indirectly above.

* [internal/core/adt: implement new evaluator (1174341)](https://review.gerrithub.io/c/cue-lang/cue/+/1174341)
* [internal/core/adt: only error if inspected node missed counters (1174576)](https://review.gerrithub.io/c/cue-lang/cue/+/1174576/3)
* [internal/core/adt: closedness improvements (1174014)](https://review.gerrithub.io/c/cue-lang/cue/+/1174014/11)
* [internal/core/adt: add lifetime check for closedNess (1174013)](https://review.gerrithub.io/c/cue-lang/cue/+/1174013/11)
* [internal/core/adt: mark functions only used in old evaluator (1174006)](https://review.gerrithub.io/c/cue-lang/cue/+/1174006/12)
* [internal/core/adt: exclude tests with disjunctions (1174004)](https://review.gerrithub.io/c/cue-lang/cue/+/1174004/13)
* [internal/core/adt: propagate closedness up (1174003)](https://review.gerrithub.io/c/cue-lang/cue/+/1174003/12)
* [internal/core/adt: introduce visual CUE debugger (1174002)](https://review.gerrithub.io/c/cue-lang/cue/+/1174002/11)

**Comment by mpvl** [comment]
Development on the new evaluator is progressing steadily. The main focus remains implementing the new evaluator as the basis for performance and other improvements ([#2884]([https://github.com/cue-lang/cue/issues/2884](https://github.com/cue-lang/cue/issues/2884))) and making the new evaluator available via a `CUE_EXPERIMENT` ([#2886]([https://github.com/cue-lang/cue/issues/2886](https://github.com/cue-lang/cue/issues/2886))). As a reminder, we track the main areas of work on the evaluator via the [Evaluator Roadmap]([https://github.com/orgs/cue-lang/projects/18](https://github.com/orgs/cue-lang/projects/18)).

At the time of our [last update](https://github.com/cue-lang/cue/issues/2850#issuecomment-1971419313), 87% of all evaluator test files were run and free of P0 or P1 bugs, while 19 files were excluded from the tests entirely. 

Now, 94% of these tests pass, with only 3 files excluded from tests. Moreover, before many files had many errors. Now, the 23 (out of 346) remaining files have at most a few errors remaining.

_(As a reminder, we categorize bugs/errors according to their severity: P0 being critical, through to P3.)_

We also added some new tests. Here is a new benchmark that we will add:

```cue
#def: {
	{} | {c: string}
	{} | {c: string}
	{} | {c: string}
	{} | {c: string}
	{} | {c: string}
	{} | {c: string}
	{} | {c: string}
	{} | {c: string}
	{} | {c: string}
	{} | {a: string}
}
x: #def
x: c: "foo"
```

The old evaluator completed this with this many core operations:

```
Unifications:  2590
Conjuncts:  13409
Disjuncts:  4674
```

Incrementally adding lines showed that this test displayed exponential behavior.

You can collect stats for your own runs using the `CUE_STATS_FILE` environment variable. This feature is subject to change.
  
The new evaluator accomplishes the same result with far fewer operations:

  
```
Unifications:  28
Conjuncts:  267
Disjuncts:  76
```

The new evaluator displays linear growth in the number of operations when adding a repeated line. Clearly an improvement.

What made this possible is the capability of the new evaluator to much more accurately detect duplicates, even when a value is only partially evaluated.  

That said, there is still some work to be done. This new detection algorithm needs tuning. Some of the cycle detection algorithms are also not fully functional yet. This means that for some of the benchmarks the new evaluator performs worse, for now. But all the techniques that worked for the old evaluator will work for the new one, in addition to all the improvements made possible by the new implementation.

There were also some complications that lead to some tweaking to the [disjunction](https://github.com/cue-lang/cue/issues/2851) algorithm. It turned out that the implementation would be much more straightforward if we had a partial implementation of [structure sharing](https://github.com/cue-lang/cue/issues/2854). We opted for this approach. This meant a slight setback in the projected time for implementing disjunctions in the new evaluator. On a positive note, however, it means we are now well on our way to implementing structure sharing. We will not need to finish this before launching the new evaluator in experimental form, but we will be able to introduce structure sharing sooner than expected. This is great, as there are some known performance issues that cannot be solved without it.

We intend to launch the [experiment](https://github.com/cue-lang/cue/issues/2886) when all known P0 and P1 bugs have been fixed. We anticipate being able to do so before the end of March 2024.

And as a reminder, this is before we improve [memory management](https://github.com/cue-lang/cue/issues/2887). The new evaluator currently has a high constant overhead. This is known and expected. We do hope, however, that people will see a big reduction in core operations.

**Comment by mpvl** [comment]
Apologies, this update is arriving a week later than planned.

We are getting close to releasing an experimental version of the [new evaluator](https://github.com/cue-lang/cue/issues/2884): we are currently at over 95% of the test files passing.

This seems like only a small increase since [last time](https://github.com/cue-lang/cue/issues/2850#issuecomment-2003925504), where we were at 94%. There is a lot going on though. The number of skipped files, the most severe of all,  went from 13 to 4.  Also, the total number of errors in the 13 remaining erroneous files has reduced as well, which is not reflected in this count. Also, the severity of the errors has diminished. Finally, we have discovered several new errors in the process that we previously forgot to count, which slowed down the reduction a bit.

In some cases we have also deliberately introduced new minor errors. For instance, we realized that some of the handling of structural cycle detection edge cases, which is not yet implemented for the new evaluator, can be handled much more nicely with [structure sharing](https://github.com/cue-lang/cue/issues/2854). We are close to supporting structure sharing. So it seemed more prudent to have a less precise, but simpler handling for now, at the expense of having some spurious structural cycles in the meantime. We can use the saved time to roll out structure sharing sooner.

We have also prepared the integration of the new evaluator into the `cue` command. So far, most of the `cue` commands, with the notable exception of `cue trim`, work with the new evaluator.

The main focus remains completing implementation of the new evaluator ([#2884]([https://github.com/cue-lang/cue/issues/2884](https://github.com/cue-lang/cue/issues/2884))) and making the new evaluator available via a `CUE_EXPERIMENT` ([#2886]([https://github.com/cue-lang/cue/issues/2886](https://github.com/cue-lang/cue/issues/2886))). There are currently [36 outstanding changes](https://review.gerrithub.io/c/cue-lang/cue/+/1191591) that need to be reviewed and submitted for the experimental evaluator to be available in tip. We will update in the coming week when that work is complete. 

Unlike what we said last time, the initial  experimental version will contain some outstanding P0 and P1 bugs. The outstanding bugs are mainly regarding structural cycle detection, most notably when used for implementing recursion, let processing, and some smaller issues. We realize that those that do not use these constructs could already benefit from the new evaluator.
 
Getting early exposure and feedback will ultimately, however, speed up the rollout.

**Comment by mpvl** [comment]
Today, we will enable the new evaluator as an experiment for the cue binary as part of v0.9.0-alpha.2. At a later point, we also plan on enabling it in the API.

The new evaluator has various known bugs that have not yet been fixed. On the flip side, it has more performance improvements than we initially planned to release.

We decided in favor of this tradeoff because, on the one hand, many users will not run into the remaining bugs and could benefit from the performance gains now, while on the other hand, some of the performance improvements will actually help address some of the remaining missing functionality that gives rise to these bugs.

Also, by enabling more users to benefit from performance gains sooner, we hope to also get more feedback on the remaining issues.

So in spite of the remaining bugs, we suggest you give the new evaluator a whirl. In our experience, it is already significant faster than the old evaluator due to some significant big-O wins.


### Using the new evaluator

To enable it on the command line, use the `evalv3` option via `CUE_EXPERIMENT`.
For instance:

```sh
CUE_EXPERIMENT=evalv3 cue export
```

Together with the `CUE_STATS_FILE=stats.json` option, you can see in terms of the number of core operations how the old and new evaluator compare. Note, though, that counts are not entirely comparable: the new evaluator overcounts conjunctions, while the old evaluator overcounts disjunctions.


## Known issues

### Structural cycles

Not all of the core constructs have been ported from the old evaluator to the new one. A noteworthy one is structural cycles. We have seen that the new algorithm gives rise to a simpler and more effective structural cycle handling. Structure sharing plays a role in that.

Instead of porting the rather complex old mechanism to the new evaluator, we have decided to implement an approximation instead for now. Together with structure sharing, this already covers most of the cases. But there are cases where structural cycles are either detected too late or spuriously.

### Let expressions in comprehensions

There are known issues with handling of let expressions in comprehensions.

### Closedness scope

There are some remaining issues in closedness detection. It is known where these come from, we just did not want to hold up the release on these issues, as we expect they will not affect most users.

This issue may manifest itself by disjuncts that are not eliminated that should otherwise be eliminated because of closedness constraints.

### Error handling

The new evaluator does not have all the locations relevant for an error as was available in the old evaluator. As soon as we are on feature parity, we plan to give a thorough overhaul on error reporting, fixing those features and others.


## Performance changes

The new evaluator is significantly faster on many fronts. In addition to a much more capable disjunction elimination algorithm, it has a partial implementaton of structural sharing, which has proven to give some significant wins.

### Gains

Both the re-implementation of disjunctions as well as structure sharing has shown significant performance improvements.

This example is similar to the hypothetical example we showed in the last update. It reflects a real-life version of it.

```
#Secret:  $secret: id: string
#secrets: #Secret | {[string]: #secrets}

out: #secrets & {
	FOO: $secret: id: "100"
	ONE: TWO: THREE: $secret: id: "123"
}

#Secret:  $secret: _id: string
#secrets: #Secret | {[string]: #secrets}

out: #secrets & {
	FOO: $secret: _id: "100"
	ONE: TWO: THREE: $secret: _id: "123"
}
```
Here we saw a reduction in the number of operations from
```
Unifications: 740771
Conjuncts:    3143640
Disjuncts:    254254
```
to
```
Unifications: 449
Conjuncts:    22643
Disjuncts:    4560
```

An example of the kind of improvements brought by structure sharing is this code:
```
f1: string
f2: f1
f3: f2
f4: f3
f5: f4
f6: f5
…
f997:  f996
f998:  f997
f999:  f998
f1000: f999
```
Where we reduce the number of conjunct operations from `500501` to `2001`, basically as a result reducing this case from `O(n^2)` to `O(n)`.

Similarly, a reduced version of the [Billion Laughs attack](https://en.wikipedia.org/wiki/Billion_laughs_attack), shown here
```
f: [g, g, g, g, g, g, g]
g: [h, h, h, h, h, h, h]
h: [i, i, i, i, i, i, i]
i: [j, j, j, j, j, j, j]
j: 1
```
reduces the number of operations from
```
Unifications: 3268
Conjuncts:    6529
```
to
```
Unifications: 34
Conjuncts:    95
```
Here the new evalautor is `O(n)` in the number of input nodes!

Finally, the improvements of the new disjunction algorithm can also compound with structure sharing. This is an example where we have recursion combined with disjunctions, a common way to get into serious performance issues with the old evaluator.
```
A: #Task
B: #steps: #Script & {mount: [A]}
C: #steps: #Script & {mount: [B]}

#Script: {mount: [...#Task]}

#Task: {
	#ref
	#ref
	_ | {}
	_ | {}
	#steps: #Script
	...
}

#ref: {a: 1} | {b: 2}
```
The old evaluator clocked in at the following number of operations:
```
Unifications: 18724
Conjuncts:    100730
Disjuncts:    5373
```
The new evaluator, with its new disjunction algorithm, only needed 
```
Unifications: 2716
Conjuncts:    48270
Disjuncts:    3272
```
A reasonable improvement, especially since it is counts conjuncts more
aggressively than the old evaluator.

Adding structure sharing in the mix, however, this further reduces to
```
Unifications: 53
Conjuncts:    431
Disjuncts:    40
```
Now we're talking!


### Known performance issues

There are a few cases where it may be slower than the old evaluator:

- The caching of the old evaluator has not been fully ported, and thus in some cases, the new evaluator may be slower.

- The constant overhead per node is much higher for the new evaluator. Basically, unlike with the old evaluator, we have not yet done any memory allocation optimizations. Also, it has some significant overhead for gathering debugging information.

- The disjunction deduplication algorithm still needes tuning. There is a good change we still miss opportunities to discard paths in disjunction processing early on. If you have any disjunction-related performance issues, please let us know so that we can further tune this algorithm.

### Unimplemented performance improvements

The new evaluator does not address all performance issues yet. Note that the old evaluator also does not implement these, so there is no benefit to stick to the old evaluator here. We mainly want to point these out on what is still in store. While by no means being exhaustive, here are some examples:

- Discriminator fields for disjunctions. There is still a lot of gain to be made making disjunctions run in linear time in many cases.
  
- Fine-grained structure sharing. We know that this will be critical for configurations with a lot of chained references that are more complex than that is covered with simple structure sharing. Especially `tools/flow` applications may suffer from this. We had a prototype for the old evaluator that showed large improvements on this front, but due to limitations of the old evaluator we were not able to release. This new evaluator makes this more feasible, and this is one of the higher-priority performance gains to roll out.

- Merging of very large structs. Merging structs is still O(n^2) in the number of fields. This is with a smallish constant overhead, but it may matter for very large structs. We know that this can be done in O(n) time. We plan to implement an O(n) algorithm along with the design of a more principled approach to field ordering.

- Memory required for closedness computation is not reclaimed and thus is still expensive. The new evaluator is designed to be able to reclaim and reuse this memory in small, compact chunks, but this has not yet been implemented. Also the tree depth of this algorithm can still be significantly reduced.


## Feedback

We would love for you to give it a spin and get some feedback. The sooner we know about issues, the sooner we can fix it. The new evaluator is not only faster, but also quite a bit easier to debug, we have found. So do not hesitate to help us get you up and running faster.

**Comment by mpvl** [comment]
Since last release, we have fixed quite a few more bugs. Based on popular demand, however, we have started on a little diversion from fixing bugs to work on integrating the new evaluator in the API.

## API work

Hooking up the new evaluator to the API may seem like a small step, but is actually quite involved. The evaluator itself is quite low level. The API provides a layer on top that interprets the low-level evaluation results and presents it to the user in a more convenient manner. It is actually good we have it, as it allows us to reimplement an evaluator without users having to change their API usage!

There are many subtle and less subtle ways in which the new evaluator changes the representation, for instance:

* pattern constraints and what fields are allowed has been completely rewritten,
* conjuncts are now hierarchical,
* disjunctions are represented differently,
* structure sharing is now allowed,
* many other smaller changes.

The API needs to take all these change into account, and now handle both.

Luckily the API itself has many tests separately from the core evaluator. We are now applying the main testing framework we developed for the evaluator to track deviations to the API as well, where possible, or coming up with other solutions where this is not the case. You can track the progress at Issue #3060.

We have already fixed many of the API issues, and work on this is progressing nicely. It even resulted in some minor bug fixes in the core evaluator itself.

## Next steps

After the biggest bugs have been removed from the API support, we plan to release an option to enable the new evaluator.

After this is done, we will resume fixing the outstanding larger bugs in the evaluator. For instance, `let` evaluations can still be rather buggy.

## Feedback

As usual, we would love to hear your feedback. Please raise issues or get in touch with us via Slack.

**Comment by mpvl** [comment]
The upcoming v0.9.0-alpha.5 release is another big step forward for the new evaluator. 

We have fixed some of the major bugs that were outstanding. Based on popular demand, however, our main focus has been to integrate the new evaluator in the Go API. This is not as trivial as it sounds. So far we have rolled out about 40 CLs to accomplish this. The main reason for this is that the underlying data representation of the new evaluator differs from the old evaluator in some important aspects. This is hidden from the user in the API, but the API had to be adjusted to accommodate this.

So far we have squashed most of the couple of dozen regressions that originated from this move and have only two relevant regressions left. Some of the remaining issues include:

* `Value.Unify`: recursive closedness is not always handled correctly for.
* when merging and outputting schemas there are some issues remaining regarding aliases
* builtins are not always evaluated when exporting
* some errors related to non-existing fields are dropped.
* `tools/trim`: mostly does not yet work with the new evaluator.

Things like dependency analysis, `tools/flow` and various other packages fully pass all tests with the new evaluator.


### How to enable the new evaluator

The new evaluator can be enabled in the Go API by passing options to `cue/cuecontext`:

```
ctx := cuecontext.New(cuecontext.EvaluatorVersion(cuecontext.EvalV3))
...
ctx.CompileString("a: 1")
```

Using `EvalV3` will ensure the version is fixed once the new evaluator stabilizes. Use `EvalExperiment` to always resort to the latest evaluator experiment.

There is also an option to pass debug flags. This takes the same form as the flags supported by the environment variables. The most notable option is the one to turn off structure sharing, which is enabled by default. If you encounter a bug that you suspect may be related to structure sharing, it can be turned off using the `CUE_DEBUG` option:

```
ctx := cuecontext.New(
	cuecontext.EvaluatorVersion(cuecontext.EvalV3),
	cuecontext.CUE_DEBUG("sharing=false"),
)
...
ctx.CompileString("a: 1")
```

Note that for the `cue` command the same can be accomplished by setting the `CUE_DEBUG` environment variable to `sharing=true`. Environment variables do not affect the API usage: all options need to be set explicitly in the Go code. To use the existing debug environment variables, use `cuecontext.CUE_DEBUG(os.Getenv("CUE_DEBUG"))`.


### Core evaluator progress

With respect to progress of the core evaluator itself, there are now no more skipped tests left, and just about a dozen of high-priority regressions remain. Now all tests are running, we can see a 83% reduction in running time of running the same tests for the old evaluator versus the new evaluator. Of course this is just a snapshot, but it gives some indication of what to expect. Of course, the new evaluator will be much faster for some cases, and will be equal in some other cases. Keep in mind that we still have a lot of optimizations in store, some quite large as well.

**Comment by mpvl** [comment]
Since the last update, we have been working on responding to user bugs as well as tackling the remaining bugs. There have been a couple of weeks worth of work on other things, including a new embed proposal, and also some investigations on other blockers, which, not entirely coincidentally, also impact the new evaluator.

We have removed most bugs of the API support for the new evaluator. There are a few left. The biggest omission is CUE trim. The old trim algorithm was fully designed around the old algorithm and its respective data structures. Consequently, as the new evaluator often uses different data structures, trim completely stopped working for the new evaluator. [Matthew Sackman](https://github.com/cuematthew) is now looking into a redesign of the trim algorithm, addressing a whole slew of bugs, while also making it work for the new evaluator.

Another aspect of the new evaluator that has greatly been improved is subsumption. Some of the improvements and bug fixes made it also in the old evaluator. But for the most part, these improvements are only supported for the new evaluator. These improvements will be crucial for both optimization and defining backwards compatibility for CUE schema.

Finally, the number of bugs of the core evaluator has been steadily decreasing as well. Moreover, we are cleaning up the algorithm of the structural cycle detection of the new evaluator to handle more cases in a simpler fashion. This was one of the major omissions of the new implementation and appears to be a factor in many of the remaining bugs.

We aim for the new evaluator to be usable for most users by the end of August.

**Comment by mpvl** [comment]
After returning from our summer break, we have picked up work on the evaluator again.  Since our last update, most of the new evaluator work has gone in to addressing user-reported bugs. We have fixed a myriad of performance bugs, semantic bugs, and panics. @cuematthew has already been helpful and is continue to ramping up on the internals of the evaluator.

Another source of `v3`-related bugs was the work related to JSON Schema. We are doing a big overhaul of CUE to significantly improve compatibility and coverage. We are ensuring that this new technology also works with `v3`, which also provided a nice stress test for the `v3` evaluator.

Other work has been focussed on reducing the number of closeContext allocations and getting the closedness algorithm right. We have not yet enabled lifetime management, which means that `v3`  still does exorbitant numbers of allocations. We are using this to expose bad performance bugs as much as possible. Our intention is still to get `v3` to be much faster than `v2` _before_ we add such optimizations.

Another noteworthy development is the development of an internal "TinyCUE". Matthew has been developing a parallel new implementation of CUE, a simpler reference implementation, in order to improve his overall understanding of CUE evaluation. We are investigating to keep maintaining this implementation to insure quality in the future.

We were hoping to be able to switch over the default to `v3` now. A few things are stopping us from doing so. One big discrepancy is field order. The difference between `v3` and `v2` is too inconsistent. We want to address this by having a better defined field order, and make field order part of the spec. We will have an offsite mid-September to sit together and speed through some of these remaining issues.

That said, we have several users increase their use of `v3` and we are ramping up the usage internally to get it to a point that it can be made the default as quickly as possible.

**Comment by mpvl** [comment]
It has ben a while since the last update. We have continued our focus on user-reported bugs. Also the JSON Schema work has exposed more bugs that we have addressed. Since last time, we have closed close to 50 issues related to the new evaluator. Part of this was designing a new and simplified cycle algorithm. The new algorithm is considerably simpler compared to the old.

New since last time is the focus on Unity. We are working to remove discrepancies in Unity. So if you want some assurance the new evaluator will work with your code, be sure to add your repo to Unity!

There are about 10 main issues left in making Unity compatible. One of them is actually a closeness bug fix that causes too much backwards incompatibility. Consider this example:
```
#Def: a: 1
x: (#Def & {b: 2}).b
```
The old evaluator would allow this, even though `#Def` does not allow field `b`. The new evaluator fixed this bug, but, unfortunately, it causes a lot of configurations to break. Our intention is to reintroduce the bug to simplify the transition to the new evaluator and fix this at a later point.

With respect to outstanding bugs, the new evaluator is now almost on par with the old one. We are considering flipping the default as soon as Unity passes and some known remaining performance issues are fixed.

With respect to Unity, @cuematthew has completed a first round on the field ordering work. The aim is to make differences between V2 and V3 easier to compare.

Aside from the remaining performance issues, we also have made some headway with fine-grained structure sharing. This has potential to greatly improve the performance of `tools/flow` as well as any configuration that has longish chains of references. We will not focus on fixing the memory management before making the switch. We expect that the outstanding algorithmic improvements will have a much bigger impact on performance than addressing the memory allocations. Of course we will still address this the future.

**Comment by mpvl** [comment]
Firstly, I wanted to wish you all a Happy New Year and hope you had a chance to have some good recovery time over the Holidays.

I also would like to thank all of you that filed issues!

Since last update, we have made again considerable progress on getting V3 ready. Let's first express this in numbers:

* GitHub issues: we still have about 10 open issues. However, there were 28 new issues added, so about the same number fixed.
* Test suite: we have now fixed **99%** of the P1 and P0 issues. There are three seemingly related P1 issues open. As a reminder, P2 and above are mostly error messages and reorderings.
* Unity: we have eliminated several key bugs here. There are still 3 projects with performance issues and 3 with unexpected errors, which are reflected by issues in GitHub.
* Tooling: @cuematthew implemented a new version of `trim` that not only makes trim work with V3, but also fixes a whole host of issues with the old version.
* Our own CI, which is an interesting test case, now runs on V3.


### Remaining performance issues

@mvdan analyzed some of the remaining performance issues. At least one of those is related to the fact that V3 does not reused helper structs, as V2 did. Basically, the number of operations is similar between the two versions, but the number of allocations goes from dozens to millions.

We have mentioned before that we intentionally did not do this as we expect that algorithmic improvements will make up for the difference, allowing us to postpone this optimization. Judging from the problem cases, as well as some constructs that are known to be slow in V3 (and V2 for that matter), we predict that most of these issue can indeed be solved with fine-grained structure sharing, the implementation of which is next on the docket.

In case there are cases left that do not benefit from this optimization, however, we have also made considerable progress on cleaning up the lifetime management of objects. So in case the fine-grained structure does not help enough, full blown memory management will be easier to implement.

**Comment by mpvl** [comment]
We are about to have our community call, which seems like a good opportunity for an update.

Thanks again for all those that filed issues!

Some highlights since the last update:

* We closed about 59 evalv3-related issues
* close to 100% of test suite issues fixed (only 1 P1 bug left, all P0 bugs closed)
* Further simplification and some important bug fixes in the structural cycle detection algorithm.
* some big performance improvements

As of now, there are 15 evalv3-related bugs outstanding. Of those, four are performance related. Most of these are related to how slight semantic changes in the API affects tools like the OpenAPI generation. Aside from that, we have 2 remaining issues left in Unity.

### Remaining performance issues

The most remaining work is related to the outstanding performance issues. We have a better understanding of what they are. We started implementing finer-grained structure sharing, which we believe can reduce the running time considerably.

We also identified that closedness computation is expensive and now drives most of the cost. We have some ideas on how to simplify closedness rules that would make them more pragmatic, but also drive down most of the cost of CUE evaluation.

We will be working on both fine-grained structure sharing as well as looking into closedness simplification in the upcoming cycle.

**Comment by mpvl** [comment]
We let our performance update coincide with our community call.


There are some big updates since last time. We now have the `evalv3` enabled by default in our v0.13 alpha. This was made possible by a bit of a pivot we made with implementation of the closedness algorithm.

As mentioned in last community call, we identified the closedness algorithm as inherently slow. This was also the case for `evalv2`. This meant, however, that there were inherent limitations on how much speedup we could expect for `evalv3`. Some of the slow configurations in Unity could not be expected to be improved by big-O optimizations in the core evaluation algorithm alone.

Around the same time, some user feedback came in that found the current closedness behavior, especially when it pertains to embeddings, unintuitive. The combination of these two factors made us realize we can actually come up with a much simpler closedness algorithm that makes things both more intuitive and faster. A quick spike showed that such an alternative algorithm would indeed speed things up considerably. More about this algorithm below.

## Status

### Github Issues

We now have 6 oustanding Issues related to `evalv3` in GitHub, down from 15 last time. In the process of making `evalv3` the default, we uncovered many Issues, that were subsequently also fixed. Three of the issues are related to the new closedness algorithm. The remaining issues are previously known performance issues that we will need to address by implementing memory management.

Issue https://cuelang.org/issues/3881, which is a performance regression related to the new closedness algorithm, deserves special mention, as it affects several of the Unity-related projects.

### Unity

With the last update, we had 12 out of 20 Unity projects passing and without performance regressions. Today this is 16 out of 20.

Two of the four remaining projects still suffer from performance regressions, which we expect to be fixed by adding memory management. Two are affected by the remaining bug in the new closedness algorithm.

Of the projects that now work on Unity, we often see significant performance improvements. One project (`tmm1/taxes.cue`) would not finish before 10 minutes and now runs in about 1 second.

The largest project in Unity used to run for 70s and now runs in 40s. This is not a huge improvement (-43%), as we are aiming for orders of magnitude.  Analysis shows that this project is also greatly affected by the closedness performance regression. So taking this into account, as well as the upcoming memory management and fine-grained structure sharing, we expect to see significant further improvements.


## Current status of closedness

After implementing the new closedness algorithm, we saw there would be quite a few incompatibilities. We realized, though, that with some adjustments we can closely approximate the `evalv2` closedness algorithm. This slows down the algorithm a bit, but it greatly simplifies the transition.

Note that most of the remaining bugs were related to the old closedness algorithm. So as a side effect, all these bugs are now fixed.

Moreover, even though there are some incompatibilities using the new algorithm, the new implementation is actually much closer to `evalv2` than the original implementation for `evalv3`. We expect the transition will be smoother for most users. In fact, we removed the `openinline` option altogether, as it is no longer needed to pass the Unity projects: they are now compatible out of the box.

A big benefit of the new closedness algorithm is that we can always eliminate if a conjunction is unified in the same position more than once. Previously, this was not the case. This especially reduces the amount of work that has to be done for disjunctions. There are some cases where we may need to do more work, but overall we have seen a significant reduction in operations. There are now no known cases where the number of operations in `evalv3` is worse than in `evalv2`, and often it is significantly better. The projects that are slower are so because of memory allocations.

We saw our core `evalv3` tests running time decrease by 50%. Note that these are mostly small tests. The effect can be expected to be larger for large configurations.

Another big benefit of the new closedness algorithm is that it greatly simplifies implementing memory management for the main algorithm as well.

We still plan to propose the changes that inspired the new closedness algorithm and allow for further simplification and speedup. We will post this as a Github discussion proposal.


## Disabling closedness checking

We have added a new `CUE_DEBUG` option to disable closedness checking altogether. This may speed up evalution speed considerably in some cases. To enable this use
```bash
CUE_DEBUG=opendef
```

One should note that a closedness violation may currently eliminate disjunctions, meaning that the result of evaluation may change, for instance by resulting in unresolved disjunctions. In the future we may move to closedness purely being a typo check and not influence disjunction elimination.

## Next steps

Our first aim is to finalize the closedness algorithm. It is considerably simpler than the old algorithm, and we already have less outstanding bugs than with the previous implementation.

There are still some incompatibilities with OpenAPI and WASM that need to be addressed as well.

Finally, with the little detour of implementing a new closedness algorithm, we have postponed the memory management and fine-grained structure sharing. We plan to take this up again soon.

**Comment by mpvl** [comment]
Since our last update, we’ve made continued progress on stabilizing and improving `evalv3`, which is now the default in `v0.13`. This marks a significant milestone in the development effort. While some issues remain, the overall maturity and performance of `evalv3` have improved substantially.

In particular, we’ve been able to address many of the remaining bugs and compatibility issues that were reported in previous cycles. The closedness algorithm introduced last round continues to prove effective and has laid the groundwork for further optimization work.

## Status

### GitHub Issues

We’ve made good progress cleaning up the remaining issues around `evalv3`. The number of closed issues has increased further since last time, and many of the remaining ones are either cosmetic or pertain to known edge cases that have low impact on typical usage. Overall, the number of bugs in `evalv3` is considerably reduced relative to those in `evalv2`.

As before, we’re tracking some regressions tied to specific large configurations, where evaluation still incurs a heavy memory load or suboptimal structure sharing. These are now our primary focus and are actively being addressed. One such issue, [#3881](https://cuelang.org/issues/3881), continues to be relevant, though the number of affected projects has decreased.

We’re also seeing that the new closedness algorithm, although still a departure from `evalv2`, is settling into a stable shape. Feedback has helped identify remaining rough edges, especially when it comes to certain embedding patterns.

### Unity

We now have **20 out of 25 Unity projects** passing both correctness and performance tests, up from 16 out of 20. This is not only a quantitative improvement but also a qualitative one, as the two newly passing projects were previously blocked by closedness incompatibilities or memory-related slowdowns.

Two of the remaining three projects are primarily limited by performance regressions, which we believe will be resolved by the upcoming work on memory management and fine-grained structure sharing. The third is affected by a subtle interaction between disjunction elimination and closedness, which we are currently experimenting with.

Performance gains continue to be substantial for many projects. One notable case improved from minutes to seconds in evaluation time. Another large configuration went from 70s to around 35s—about a 50% reduction. While not yet “orders of magnitude,” this is encouraging, especially given that further improvements are anticipated from structural sharing enhancements currently in development.

## Current Status of Closedness

The simplified closedness algorithm continues to hold up well under broader usage. In practice, its behavior has aligned closely enough with `evalv2` to make the transition smooth for most users. We have not seen new classes of user-facing bugs arise from this change, and in fact, many prior bugs have been resolved by moving to the new implementation.

Because of the new approach, we’ve been able to eliminate several previously complex heuristics. In particular, we can now avoid unnecessary recomputation in disjunction handling, leading to both a cleaner implementation and better runtime characteristics.

We are currently working on a refinement of this new algorithm that should fix the remaining edge cases, as well as further improve performance.

## Disabling Closedness Checking

As before, developers may opt out of closedness checking entirely using the environment variable:

```bash
CUE_DEBUG=opendef
```

This remains useful for performance debugging and understanding the impact of closedness constraints on evaluation. But we recommend using it with caution, as it may change evaluation outcomes in some cases, especially those involving disjunctions.

## Next Steps

Our focus is now squarely on memory management and fine-grained structure sharing—two areas that promise significant further improvements in evaluation speed and resource usage. With the closedness refactor largely behind us, we can devote more resources to these foundational optimizations.

We are also reviewing potential simplifications to the core evaluation loop made possible by the newer closedness model. These could further reduce runtime overhead and pave the way for a cleaner internal architecture.

Finally, we plan to continue reducing the remaining Unity failures and will aim to bring the success rate closer to 100% in the next release cycle.

**Comment by mpvl** [comment]
It has been a while since we posted our last performance update. We have shifted from focussing on specific known algorithmic improvements to focussing on specific issues reported by users and our Unity projects. We have now gotten all our Unity projects without reasonable bounds (and many much faster). 

## Bye bye `evalv2`

As now all Unity projects are within bounds, and most known bugs have been eliminated (there are far more bugs in evalv2), we have now taken the step to remove `evalv2` in its entirety.

## More closedness improvements

We mentioned before that we identified closedness as being the main issue. We have done another revision of the closedness algorithm, giving another 90% speedup for several projects. The core principle remained mainly the same, but we reduced the basic algorithm from being polynomial to linear times the depth of the "closedness tree". 

The closedness tree can still be fairly large for in the presence of many nested embeddings. The new algorithm is also designed to work well with the new `explicitopen` experiment (see Proposal #4032), which greatly reduces this depth. So expect more speedups in the future as we shift to this new semantics.

## Flow, Syntax, and export

We recently also discovered a performance bug in the dependency analyzer. This is not part of the core evaluator, but affects many subsystems, such as `cue export`, `Syntax`, and the `flow` package. The issue was that the same dependency could be analyzed many times over. We have now reduced this exponential behavior to linear. In one case, this reduced running time from `47s` to `0.4s`.

## Future work

There are many possible performance improvements that are still open:
- fine-grained structure sharing
- reduce allocations of `Environment`
- improved memory management
- discriminator fields for disjunctions
- etc.

So far we have been able to get where we are without touching those. Each of these can still significantly improve performance.

We have also started looking into "`evalv4`", which is really a new, precise, scheduler for `evalv3`. This has the potential of speeding up comprehensions. But it would also incorporate external functions natively into the evaluation model, which means we could run CUE workflows without the need for dependency analysis and purely rely on CUE's dataflow nature.

**Comment by mvdan** [comment]
I've done quite a bit of performance work over the last two months, mainly in terms of avoiding allocations to reduce GC overhead and memory usage, but also other localized optimizations as measured by CPU profiles.

See our [last community update](https://github.com/cue-lang/cue/discussions/4193) thread for more info; the slides and recording show how I briefly presented these results. Multiple Unity projects saw speed-ups of between 15% and 80% when comparing master with v0.15.0, depending on why exactly each of them was being slow.

At this point we expect practically every project to be faster with evalv3 on master than with the old evalv2. If you have a project where this is not the case, please open a new issue and we can investigate.
